app/battle/knight.py

```diff
@@ -0,0 +1,44 @@
+class Knight:
+    def __init__(self,
+                 name: str,
+                 power: int,
+                 hp: int,
+                 armour: list,
+                 weapon: dict,
+                 potion: dict,
+                 ) -> None:
+
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.protection = 0  # Default protection
+
+        self.apply_armour(armour)
+        self.apply_weapon(weapon)
+        self.apply_potion(potion)
+
+    def __str__(self) -> str:
+        return self.name
+
+    def apply_armour(self, armour: list) -> None:
+        for piece in armour:
+            self.protection += piece.get("protection", 0)
+
+    def apply_weapon(self, weapon: dict) -> None:
+        self.power += weapon.get("power", 0)
+
+    def apply_potion(self, potion: dict) -> None:
+        if potion is not None:
+            self.hp += potion.get("effect", {}).get("hp", 0)
+            self.power += potion.get("effect", {}).get("power", 0)
+            self.protection += potion.get("effect", {}).get("protection", 0)

```

Comment:
Follow DRY principle and do not repeat similar code, perform an operation in a for loop iteration over a list of `"hp"`, `"power"` and `"protection"` and using `setattr` and `getattr` functions to dynamically define arguments